### ISSUE TITLE
Incorrect dependencies reordering

### ISSUE THREAD

**Issue by fcartegnie** [pkgconf/pkgconf#361]
I have an ordering regression with link orders.

It was correct until 1.9.5 (fedora 39)
Now broken on 2.1.1 (fedora 40)

With ffmpeg >= 6.1, libswresample is listed as dependency for libavcodec

```
Name: libavcodec
Description: FFmpeg codec library
Version: 61.3.100
Requires: libswresample >= 5.1.100, libavutil >= 59.8.100
Requires.private: 
Conflicts:
Libs: -L${libdir}  -lavcodec -pthread -lm -latomic -llzma -lz -lva
Libs.private: 
Cflags: -I${includedir}
```

And libavformat itself depends on both

```
Name: libavformat
Description: FFmpeg container format library
Version: 61.1.100
Requires: libavcodec >= 61.3.100, libswresample >= 5.1.100, libavutil >= 59.8.100
Requires.private: 
Conflicts:
Libs: -L${libdir}  -lavformat -lm -latomic -lbz2 -lz
Libs.private: 
Cflags: -I${includedir}
```

When requesting libs for libavformat, lists libavcodec & swresample dependencies in the right order:

```
/usr/bin/pkg-config --static --libs  'libavformat >= 58.29.100 '

-L/lib -lavformat -lbz2 -lavcodec -lm -llzma -lz -lswresample -lavutil -pthread -lva-drm -lva-x11 -lvdpau -lm -ldrm -lva -latomic -lX11
```

But adding libavcodec does not move its own swresample dependency, which is listed in Requires:

```
/usr/bin/pkg-config --static --libs  'libavformat >= 58.29.100 libavcodec '

-L/lib -lavformat -lbz2 -lswresample -lavutil -lva-drm -lva-x11 -lvdpau -ldrm -lX11 -lavcodec -pthread -lm -latomic -llzma -lz -lva
```

Because swresample is now before avcodec, it ends with unresolved symbols from libswresample

Our reference https://code.videolan.org/videolan/vlc/-/issues/28697

**Comment by dg0yt** [comment]
Well, use 2.2.0. It should have neither the lib order nor the performance regression.

**Comment by fcartegnie** [comment]
confirmed fixed in 2.2.0 after building new system packages

**Comment by AdamWill** [comment]
The problem is that 2.2.0 comes with an ABI change and an soname bump, which is not something downstreams like to put in stable release updates. See https://bugzilla.redhat.com/show_bug.cgi?id=2307038 .

As I wrote there, it seems like the 'safest' fix for this on 2.1.1 is to revert https://github.com/pkgconf/pkgconf/pull/339 , which is what broke it, and seems to be only a performance improvement, not a bugfix. Would it be possible to do a 2.1.2 with #339 reverted ? That would be appropriate for downstreams that are on the 2.1 series to ship.
